### PR TITLE
Feat: Add support for qualified names/constants in the pattern matching

### LINKED ISSUE #5607: Qualified constants cannot be referenced inside bit array size segment patterns.
When pattern matching on a bit array, the compiler produces a syntax error if a qualified imported constant is referenced inside a size segment, for example:

```gleam
import example/internal

const lap_data_size = 50

pub fn parse(bits:BitArray)->Result(_,Nil){
  case bits {
    <<
      lap_data:bytes-size(internal.maximum_cars * lap_data_size),
      time_trial_personal_best_car:unsigned-int-8,
      time_trial_rival_car:unsigned-int-8,
    >> -> todo
    _ -> todo
  }
}
```

This code produces the following error pointing to `internal.`:

```
Syntax error

I was not expecting this.

Found `.`, expected one of: 
- `)`
```

### PR DESCRIPTION
**Pull request by 13dev**
In this feature was added a module to the AST, so we can parse the qualified names in the variables.

This closes #5607 .

- [x] The changes in this PR have been discussed beforehand in an issue
- [x] The issue for this PR has been linked
- [x] Tests have been added for new behaviour
- [x] The changelog has been updated for any user-facing changes

### CODE REVIEW

File: compiler-core/src/type_/pattern.rs [hunk 1]
```diff
@@ -1348,63 +1348,107 @@ impl<'a, 'b> PatternTyper<'a, 'b> {
...
-                        None => {
-                            return Err(Error::UnknownVariable {
+            BitArraySize::Variable {
+                name,
+                module,
+                location,
+                ..
+            } => {
+                let (constructor, reference_kind) = match &module {
+                    // look up Qualified name in the imported module
```
Comment (ankddev) [inline compiler-core/src/type_/pattern.rs]:
```suggestion
                    // Look up qualified name in the imported module
```

File: compiler-core/src/parse.rs [hunk 2]
```diff
@@ -3869,12 +3869,32 @@ where
...
     fn parse_bit_array_size_unit(&mut self) -> Result<BitArraySize<()>, ParseError> {
         match self.next_tok() {
-            Some((start, Token::Name { name }, end)) => Ok(BitArraySize::Variable {
-                location: SrcSpan { start, end },
-                name,
-                constructor: None,
-                type_: (),
-            }),
+            Some((start, Token::Name { name }, name_end)) => {
+                // Check if it has a qualified name ex:module.name
```
Comment (ankddev) [inline compiler-core/src/parse.rs]:
```suggestion
                // Check if it has a qualified name like module.name
```

File: compiler-core/src/erlang/tests/bit_arrays.rs [hunk 3]
```diff
@@ -526,3 +526,35 @@ pub fn go(wibble: String) {
 "#
     )
 }
+
+#[test]
```
Comment (ankddev) [inline compiler-core/src/erlang/tests/bit_arrays.rs:530]: Could you add similar tests to JS target too?

File: compiler-core/src/type_/pattern.rs [hunk 4]
```diff
@@ -1348,63 +1348,107 @@ impl<'a, 'b> PatternTyper<'a, 'b> {
...
+                            Some(variable) if variable.in_scope() => {
+                                variable.usage = Usage::UsedInPattern;
+                                ValueConstructor::local_variable(
+                                    variable.location,
+                                    variable.origin.clone(),
+                                    variable.type_.clone(),
+                                )
+                            }
+                            // Otherwise, we check the local scope.
+                            Some(_) | None => match self.environment.get_variable(&name) {
```
Comment (ankddev) [inline compiler-core/src/type_/pattern.rs]:
```suggestion
                            _ => match self.environment.get_variable(&name) {
```

File: compiler-core/src/type_/pattern.rs [hunk 5]
```diff
@@ -1348,63 +1348,107 @@ impl<'a, 'b> PatternTyper<'a, 'b> {
...
+                        (constructor, ReferenceKind::Unqualified)
+                    }
                 };
 
                 match &constructor.variant {
                     ValueConstructorVariant::LocalVariable { .. } => (),
                     ValueConstructorVariant::ModuleConstant {
                         name: canonical,
-                        module,
+                        module: constructor_module,
```
Comment (ankddev) [inline compiler-core/src/type_/pattern.rs]: nit: do we need to rename it if we can just shadow original variable?

File: compiler-core/src/parse.rs [hunk 6]
```diff
@@ -3869,12 +3869,32 @@ where
...
+                if let Some((_, Token::Dot, _)) = self.tok0 {
+                    self.advance();
+                    match self.next_tok() {
+                        Some((_, Token::Name { name: var_name }, end)) => {
```
Comment (ankddev) [inline compiler-core/src/parse.rs]: I think that it's better to avoid abbreviations, though idk if this is the case, since it seems to be many such abbreviations in the parser, unlike other parts of the codebase

File: compiler-core/src/parse/snapshots/gleam_core__parse__tests__correct_precedence_in_pattern_size.snap [hunk 7]
```diff
@@ -81,6 +81,7 @@ expression: "let assert <<size, payload:size(size + 2 * 8)>> = <<>>"
                                                 start: 32,
                                                 end: 36,
                                             },
+                                            module: None,
```
Comment (ankddev) [inline compiler-core/src/parse/snapshots/gleam_core__parse__tests__correct_precedence_in_pattern_size.snap:84]: Could you add parser tests for case with qualified name?

**Review by ankddev — COMMENTED** [review]
Hello! Nice addition and excellent work! I've left a bunch of notes inline.

Also, could you add a changelog entry, since it's user-facing change? And could you chnage names in tests to be more consistent with the rest (e.g. we use a lot of `wibble` and `wobble` in the compiler)?

File: compiler-core/src/type_/pattern.rs (hunk 5 above)
Comment (13dev) [reply compiler-core/src/type_/pattern.rs]: fixed :)

File: compiler-core/src/parse/snapshots/gleam_core__parse__tests__correct_precedence_in_pattern_size.snap (hunk 7 above)
Comment (13dev) [reply compiler-core/src/parse/snapshots/gleam_core__parse__tests__correct_precedence_in_pattern_size.snap:84]: Could you review the tests? thanks

File: compiler-core/src/parse.rs (hunk 6 above)
Comment (13dev) [reply compiler-core/src/parse.rs]: I clean up this part, use some available functions and make the code much smaller, can you give me some feedback about it?

File: compiler-core/src/javascript/tests/snapshots/gleam_core__javascript__tests__bit_arrays__qualified_constant_in_pattern_size.snap [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+----- COMPILED JAVASCRIPT
+import { makeError, bitArraySliceToInt } from "../gleam.mjs";
+import * as $sizes from "../sizes.mjs";
+
+const FILEPATH = "src/module.gleam";
+
+export function go(data) {
+  let value;
+  if (byte_size >= 0 && data.bitSize === byte_size) {
```
Comment (ankddev) [inline compiler-core/src/javascript/tests/snapshots/gleam_core__javascript__tests__bit_arrays__qualified_constant_in_pattern_size.snap]: `byte_size` is undefined here, so it doesn't work!

File: compiler-core/src/javascript/tests/snapshots/gleam_core__javascript__tests__bit_arrays__qualified_constant_in_pattern_size_expression.snap [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+import { makeError, bitArraySliceToInt } from "../gleam.mjs";
+import * as $sizes from "../sizes.mjs";
+
+const FILEPATH = "src/module.gleam";
+
+const local_size = 4;
+
+export function go(data) {
+  let value;
+  if (multiplier * local_size >= 0 && data.bitSize === multiplier * local_size) {
```
Comment (ankddev) [inline compiler-core/src/javascript/tests/snapshots/gleam_core__javascript__tests__bit_arrays__qualified_constant_in_pattern_size_expression.snap]: And `multiplier` here too!

File: CHANGELOG.md [hunk 10]
```diff
@@ -421,3 +421,5 @@
   offset 0 instead of the slice's actual byte offset, causing sliced bit arrays
   to read from the wrong position in the underlying buffer on JavaScript.
   ([John Downey](https://github.com/jtdowney))
+
+- Fix the bug where qualified constants cannot be referenced inside a bit array size segment pattern. ([Leo Oliveira](https://github.com/13dev))
```
Comment (ankddev) [inline CHANGELOG.md]: Wrap at 80 symbols and place your name on new line, please
```suggestion
- Fix the bug where qualified constants cannot be referenced inside a bit array
  size segment pattern.
  ([Leo Oliveira](https://github.com/13dev))
```

File: compiler-core/src/format.rs [hunk 11]
```diff
@@ -2554,7 +2554,10 @@ impl<'comments> Formatter<'comments> {
     fn bit_array_size<'a>(&mut self, size: &'a BitArraySize<()>) -> Document<'a> {
         match size {
             BitArraySize::Int { value, .. } => self.int(value),
-            BitArraySize::Variable { name, .. } => name.to_doc(),
+            BitArraySize::Variable { module, name, .. } => match module {
+                Some((module_name, _)) => module_name.to_doc().append(".").append(name.as_str()),
```
Comment (ankddev) [inline compiler-core/src/format.rs]: This conversion is useless, isn't it?

```suggestion
                Some((module_name, _)) => module_name.to_doc().append(".").append(name),
```

File: compiler-core/src/ast/visit.rs [hunk 12]
```diff
@@ -2049,6 +2049,7 @@ where
             name,
             constructor,
             type_,
+            ..
```
Comment (ankddev) [inline compiler-core/src/ast/visit.rs:2052]: I think that you should add module argument to the visitor and update all usages too! This will likely need some extra work around visiting, feel free to ask if you will need any help with this!

**Review by ankddev — COMMENTED** [review]
Looks much better! I've left some other comments, also rustfmt shows that code is unformatted in some places, could you run `cargo fmt`?

File: compiler-core/src/parse/snapshots/gleam_core__parse__tests__correct_precedence_in_pattern_size.snap (hunk 7 above)
Comment (ankddev) [reply compiler-core/src/parse/snapshots/gleam_core__parse__tests__correct_precedence_in_pattern_size.snap:84]: They look good to me, thank you!

File: compiler-core/src/parse.rs (hunk 6 above)
Comment (ankddev) [reply compiler-core/src/parse.rs]: Looks good to me!

File: compiler-core/src/javascript/tests/snapshots/gleam_core__javascript__tests__bit_arrays__qualified_constant_in_pattern_size.snap (hunk 8 above)
Comment (13dev) [reply compiler-core/src/javascript/tests/snapshots/gleam_core__javascript__tests__bit_arrays__qualified_constant_in_pattern_size.snap]: Should be $sizes.byte_size, maybe i need to update the decision.rs in compile-core javascript, right?

Comment (13dev) [reply compiler-core/src/javascript/tests/snapshots/gleam_core__javascript__tests__bit_arrays__qualified_constant_in_pattern_size.snap]: Is there fast way to test compiled JavaScript so i can catch the runtime errors easily?

Comment (GearsDatapacks) [reply compiler-core/src/javascript/tests/snapshots/gleam_core__javascript__tests__bit_arrays__qualified_constant_in_pattern_size.snap]: If you want to test the behaviour of generated code, you can add some tests in `test/language`.

File: compiler-core/src/ast/visit.rs (hunk 12 above)
Comment (13dev) [reply compiler-core/src/ast/visit.rs:2052]: maybe it is better to make this in reference at this issue? https://github.com/gleam-lang/gleam/pull/5596

Comment (ankddev) [reply compiler-core/src/ast/visit.rs:2052]: To be honest, I don't know, maybe wait for Louis?

File: CHANGELOG.md [hunk 13]
```diff
@@ -421,3 +421,7 @@
   offset 0 instead of the slice's actual byte offset, causing sliced bit arrays
   to read from the wrong position in the underlying buffer on JavaScript.
   ([John Downey](https://github.com/jtdowney))
+
```
Comment (ankddev) [inline CHANGELOG.md]: Oh, I noticed that you added entry to old changelog! you should add it to the `Unreleased` section

File: compiler-core/src/exhaustiveness.rs [hunk 14]
```diff
@@ -1965,14 +1965,19 @@ pub enum VariableUsage {
     /// A bit array named segment that was brought into scope in the bit array
     /// pattern itself and might be referenced by later segments.
     PatternSegment(EcoString, ReadAction),
-    /// A variable defined somewhere else
-    OutsideVariable(EcoString),
+    /// A variable defined somewhere else. The optional module is the name of
+    /// the module the variable is defined in if it is a qualified reference.
```
Comment (ankddev) [inline compiler-core/src/exhaustiveness.rs]: I think that it will read better with comma added
```suggestion
    /// A variable defined somewhere else. The optional module is the name of
    /// the module the variable is defined in, if it is a qualified reference.
```

**Review by ankddev — COMMENTED** [review]
I've left some other comments inline. Regarding changing visitor, it's better to ask Louis, I think

File: CHANGELOG.md (hunk 13 above)
Comment (13dev) [reply CHANGELOG.md]: oh nice catch, thanks!

Comment (13dev) [reply CHANGELOG.md]: shoot, i did a rebase and it duplicated the commits, sorry about that, CHANGELOG has changed, can you verify it?

Comment (ankddev) [reply CHANGELOG.md]: Now it looks like you added some unwanted changes (header and ending newline, look at diff), and you should place entry not only under Unreleased section, but under Bug fixes section too (just create it, like below in the file)

Comment (ankddev) [reply CHANGELOG.md]: Now it looks like you added some unwanted changes (header and ending newline, look at diff), and you should place entry not only under Unreleased section, but under Bug fixes section too (just create it, like below in the file)

File: compiler-core/src/ast/visit.rs (hunk 12 above)
Comment (lpil) [reply compiler-core/src/ast/visit.rs]: Yup, any fields need to be visited 🙏

File: compiler-core/src/parse/snapshots/gleam_core__parse__tests__correct_precedence_in_pattern_size.snap (hunk 7 above)
Comment (lpil) [reply compiler-core/src/parse/snapshots/gleam_core__parse__tests__correct_precedence_in_pattern_size.snap]: No need for parser test in future, they don't provide any particular value currently. The output is largely meaningless to human readers, and the other tests already verify that it parses as intended.

File: compiler-core/src/type_/pattern.rs [hunk 15]
```diff
@@ -1348,39 +1348,85 @@ impl<'a, 'b> PatternTyper<'a, 'b> {
...
-                                location,
-                                name: name.clone(),
-                                variables: self.environment.local_value_names(),
-                                discarded_location: self
-                                    .environment
-                                    .discarded_names
-                                    .get(&eco_format!("_{name}"))
-                                    .cloned(),
-                                type_with_name_in_scope: self
+            BitArraySize::Variable {
```
Comment (lpil) [inline compiler-core/src/type_/pattern.rs:1355]: Could you split this into two clauses please, the original one which will now match on `module: None`, and a new one with the new logic. This will make it easier to read and review 🙏

File: CHANGELOG.md [hunk 16]
```diff
@@ -1,5 +1,13 @@
 # Changelog
 
+## Unreleased
+
+### Bug fixes
+
+- Fix the bug where qualified constants cannot be referenced inside a bit array
+  size segment pattern.
+  ([Leo Oliveira](https://github.com/13dev))
```
Comment (lpil) [inline CHANGELOG.md:9]: This is a new feature, it is not a bug!

**Review by lpil — COMMENTED** [review]
Thank you! The code looks very nice.

I've left some comments inline, and could you add these tests:
- A type test that shows the error for when the value is not an int
- A language server rename test that shows it is included with renaming

Thank you

**Comment by 13dev** [comment]
Hi, sorry for the delay, can anyone make a review of the updated changes?

File: compiler-core/src/parse/snapshots/gleam_core__parse__tests__parsing_bit_array_expression_with_invalid_size.snap [hunk 17]
```diff
@@ -9,10 +9,10 @@ pub fn wibble() {
...
+  ┌─ /src/parse/error.gleam:2:29
   │
 2 │   let assert <<1:size(Upname)>> = todo
-  │                       ^^^^^^ I was not expecting this
+  │                             ^ I was not expecting this
 
-Found a name, expected one of: 
-- A variable name or an int
+Found `)`, expected one of: 
+- A variable name or an integer
```
Comment (lpil) [inline compiler-core/src/parse/snapshots/gleam_core__parse__tests__parsing_bit_array_expression_with_invalid_size.snap:18]: This error doesn't seem quite right. It says that a variable name or an int should come after `Upname`, but that's not valid Gleam.

Also we always use the name "int", never "integer".

File: compiler-core/src/type_/pattern.rs [hunk 18]
```diff
@@ -1423,6 +1427,89 @@ impl<'a, 'b> PatternTyper<'a, 'b> {
...
+                            value_constructors: imported_module.public_value_names(),
+                            type_with_same_name: imported_module
+                                .get_public_type(&name)
+                                .is_some(),
+                            context: ModuleValueUsageContext::ModuleAccess,
+                        }
+                    })?;
+
+                // Calculate the location of just the name part (after "module.")
+                let name_location = SrcSpan::new(module_location.end + 1, location.end);
```
Comment (lpil) [inline compiler-core/src/type_/pattern.rs:1476]: Could you add some tests to verify that this works correctly when there is whitespace around the `.` please

File: language-server/src/tests/rename.rs [hunk 19]
```diff
@@ -2424,3 +2424,20 @@ pub fn main() {
...
+
+pub fn main() {
+  let assert <<value:size(sizes.byte_size)>> = <<1>>
+  value
+}
+",
+        "bit_size",
+        find_position_of("sizes.byte_size").under_char('b')
+    );
+}
```
Comment (lpil) [inline language-server/src/tests/rename.rs:2443]: Could you include a test for renaming of the module at the import, and for renaming the module by moving the file please 🙏

File: language-server/src/reference.rs [hunk 20]
```diff
@@ -104,9 +105,59 @@ pub fn reference_for_ast_node(
...
+                | ValueConstructorVariant::ModuleFn {
+                    module: def_module,
+                    name: canonical_name,
+                    ..
+                } => {
+                    // For qualified references, calculate the name location by
+                    // skipping past the module and the dot
+                    let name_location = match module {
+                        Some((_, module_location)) => {
+                            SrcSpan::new(module_location.end + 1, location.end)
```
Comment (lpil) [inline language-server/src/reference.rs:122]: Same for here, could you have tests for when there is whitespace around the `.` please

File: language-server/src/reference.rs [hunk 21]
```diff
@@ -104,9 +105,59 @@ pub fn reference_for_ast_node(
...
+                    })
+                }
+                ValueConstructorVariant::Record {
+                    module: def_module,
+                    name: canonical_name,
+                    ..
+                } => {
+                    let name_location = match module {
+                        Some((_, module_location)) => {
+                            SrcSpan::new(module_location.end + 1, location.end)
```
Comment (lpil) [inline language-server/src/reference.rs:145]: As above, whitespace dot tests please 🙏

File: language-server/src/reference.rs [hunk 22]
```diff
@@ -104,9 +105,59 @@ pub fn reference_for_ast_node(
                     origin: Some(origin.clone()),
                     name: name.clone(),
                 }),
-                ValueConstructorVariant::ModuleConstant { .. }
-                | ValueConstructorVariant::ModuleFn { .. }
-                | ValueConstructorVariant::Record { .. } => None,
+                ValueConstructorVariant::ModuleConstant {
+                    module: def_module,
```
Comment (lpil) [inline language-server/src/reference.rs:109]: What's a def? No abbreviations please, they're hard to understand.

**Review by lpil — COMMENTED** [review]
Lovely work!! Thank you very much. I've added some small notes inline
